src/applications/personalization/profile360/components/PaymentInformationEditModal.jsx

```diff
@@ -0,0 +1,228 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import Modal from '@department-of-veterans-affairs/formation-react/Modal';
+import AlertBox from '@department-of-veterans-affairs/formation-react/AlertBox';
+import ErrorableTextInput from '@department-of-veterans-affairs/formation-react/ErrorableTextInput';
+import ErrorableSelect from '@department-of-veterans-affairs/formation-react/ErrorableSelect';
+
+const ACCOUNT_TYPES_OPTIONS = {
+  checking: 'Checking',
+  savings: 'Savings',
+};
+
+class PaymentInformationEditModal extends React.Component {
+  static propTypes = {
+    onClose: PropTypes.func.isRequired,
+    onSubmit: PropTypes.func.isRequired,
+    setPaymentInformationUiState: PropTypes.func.isRequired,
+    paymentInformationUiState: PropTypes.object,
+  };
+
+  static defaultEditModalFields = {
+    financialInstitutionRoutingNumber: {
+      field: {
+        value: '',
+        dirty: false,
+      },
+    },
+    accountNumber: {
+      field: {
+        value: '',
+        dirty: false,
+      },
+    },
+    accountType: {
+      options: Object.values(ACCOUNT_TYPES_OPTIONS),
+      value: {
+        value: ACCOUNT_TYPES_OPTIONS.checking,
+        dirty: false,
+      },
+    },
+  };
+
+  componentDidMount() {
+    this.setEditedState(PaymentInformationEditModal.defaultEditModalFields);
+  }
+
+  onClose = () => {
+    this.props.onClose();
+    this.setEditedState(PaymentInformationEditModal.defaultEditModalFields);
+  };
+
+  onSubmit = event => {
+    event.preventDefault();
+
+    const editedState = this.props.paymentInformationUiState.editModalFields;
+
+    const routingNumberErrorMessage = this.getRoutingNumberErrorMessage(
+      editedState.financialInstitutionRoutingNumber.field.value,
+    );
+    const accountNumberErrorMessage = this.getAccountNumberErrorMessage(
+      editedState.accountNumber.field.value,
+    );
+
+    if (routingNumberErrorMessage || accountNumberErrorMessage) {
+      this.setEditedState({
+        financialInstitutionRoutingNumber: {
+          ...editedState.financialInstitutionRoutingNumber,
+          errorMessage: routingNumberErrorMessage,
+        },
+        accountNumber: {
+          ...editedState.accountNumber,
+          errorMessage: accountNumberErrorMessage,
+        },
+      });
+    } else {
+      this.props.onSubmit({
+        financialInstitutionName: 'Hidden form field',
+        financialInstitutionRoutingNumber:
+          editedState.financialInstitutionRoutingNumber.field.value,
+        accountNumber: editedState.accountNumber.field.value,
+        accountType: editedState.accountType.value.value,
+      });
+    }
+  };
+
+  onRoutingNumberChanged = field => {
```

Comment:
I think this method is an example of where we sometimes don't take advantage of Redux. This is all data/state logic that you'd expect to see in a reducer in an app that's using Redux. In there, this logic would be plain JS that's pretty easy to test (and easier to share between multiple fields). And this component would get simpler and be focused on displaying the right information and dispatching the right action depending on user input.

Reply:
agreed that this should be put somewhere where it's easy to test in insolation. but still feeling like just a beginner with Redux, I'm not sure how putting this logic in the reducer would work. I'm guessing if I dug deep enough into the form system that question would be answered since that must be how the `ui:validations` work.

Reply:
It is, but there's a lot of extra logic around it in the forms system, so it'll be hard to figure out that way.

But essentially, you would have the React component fire an action with the info you need:

```js
onRoutingNumberChanged = field => {
  this.props.updateFormField('financialInstitutionRoutingNumber, field);
}
```
In your action creator file:

```js
function updateFormField(fieldName, field) {
  return {
    type: UPDATE_FORM_FIELD,
    fieldName,
    field
  }
}
```

And then in the reducer:

```js
case UPDATE_FORM_FIELD:
  // set from our data utilities
  return set(`paymentInformationUiState.${action.fieldName}`, {
    value: {
      ...action.field,
      // could also theoretically leave this in component, but you could write a function like this as well
      errorMessage: getErrorMessage(action.fieldName, action.field.value),
    }
  };
```

Reply:
Yeah, this is my first loop back to the React world and maybe this is standardized now, but I expected criticism if I derived these values in the reducer, because they could be considered side effects. I would prefer to have these in the reducer though.

Reply:
Thanks for that example, @jbalboni. Seeing that the `value` of each form field is an object that wraps up the "real" value along with an error message is both a "duh" and "aha" realization for me. Would the `getErrorMessage()` helper also be responsible for only validating if the field was dirty?

Reply:
That works for me @jbalboni!

Reply:
I don't think anything in here would be considered side effects. Having side effects would mean changing something other than the Redux state in a reducer, which isn't the case. The other way you could violate a "rule" in Redux reducers is to have them use some code that isn't solely dependent on the arguments passed into the reducer (i.e. use moment to get today's date). That's not happening here either, from what I saw.

Sometimes people are also strict about making the state held in Redux very minimal, but there's no hard and fast rule there.

Reply:
@erikphansen It could, but actually, I prefer always running the validation and then having the React components determine if a validation message is shown based on that dirty field. The forms system does it that way (though it doesn't keep validation state in Redux at all).

Reply:
I moved the form values into the reducer, while validation happens in the component at during  `onValueChange` . LMK if it still needs improvement though.